src/main/java/com/amazonaws/encryptionsdk/model/DecryptionMaterials.java

```diff
@@ -37,24 +78,47 @@ public Builder toBuilder() {
     }
 
     public static final class Builder {
-        private DataKey<?> dataKey;
+        private CryptoAlgorithm algorithm;
+        private SecretKey cleartextDataKey;
+        private MasterKey<?> masterKey;
         private PublicKey trailingSignatureKey;
+        private Map<String, String> encryptionContext;
         private KeyringTrace keyringTrace;
 
         private Builder(DecryptionMaterials result) {
-            this.dataKey = result.getDataKey();
+            this.algorithm = result.getAlgorithm();
+            this.cleartextDataKey = result.getCleartextDataKey();
+            this.masterKey = result.getMasterKey();
             this.trailingSignatureKey = result.getTrailingSignatureKey();
             this.keyringTrace = result.getKeyringTrace();
         }
 
         private Builder() {}
 
-        public DataKey<?> getDataKey() {
-            return dataKey;
+        public CryptoAlgorithm getAlgorithm() {
+            return algorithm;
+        }
+
+        public Builder setAlgorithm(CryptoAlgorithm algorithm) {
+            this.algorithm = algorithm;
+            return this;
+        }
+
+        public SecretKey getCleartextDataKey() {
+            return cleartextDataKey;
         }
 
-        public Builder setDataKey(DataKey<?> dataKey) {
-            this.dataKey = dataKey;
+        public Builder setCleartextDataKey(SecretKey cleartextDataKey) {
```

Comment:
the other implementations call this the 'unencrypted data key' and the spec uses plaintext.  I would pick one of these :)

At this point, the materials should verify that the data key conforms to algorithm specification.
The encrypt materials make sure that this can not be set more than once.
I would do that here as well.

These set methods generally also build the Keyring Trace.
This makes it harder to return materials with incorrect keyring traces.

Reply:
> the other implementations call this the 'unencrypted data key' and the spec uses plaintext. I would pick one of these :)

I'll go with plaintextDataKey to match the spec

>At this point, the materials should verify that the data key conforms to algorithm specification.

What do you mean by "At this point"? During the construction of DecryptionMaterials? The plaintextDataKey would be null at this point, it is set in onDecrypt in the Keyring, which uses the Algorithm in the DecryptionMaterials to create the plaintextDataKey:

```java
decryptionMaterials.setCleartextDataKey(
                            new SecretKeySpec(decryptedKey, decryptionMaterials.getAlgorithm().getDataKeyAlgo()));
                    traceOnDecrypt(decryptionMaterials.getKeyringTrace());
```

>The encrypt materials make sure that this can not be set more than once.
I would do that here as well.

This is just the builder, the actual setter for setCleartextDataKey throws an IllegalStateException if it has already been set. 

>These set methods generally also build the Keyring Trace.
This makes it harder to return materials with incorrect keyring traces.

I went back and forth on that, I wasn't happy with how it turned out when I did it that way, but let me see what I can do

Reply:
+1 on using plaintextDataKey.

> The plaintextDataKey would be null at this point, it is set in onDecrypt in the Keyring, which uses the Algorithm in the DecryptionMaterials to create the plaintextDataKey:

I think that we want to ensure that anyone who builds DecryptionMaterials, if they set a plaintextDataKey, can only successfully do so if the plaintextDataKey being set matches the algorithmSuite already set in the decryptionMaterials. i.e. let's make it harder for custom keyrings to create and return a set of materials which are not valid.

Reply:
Agreed, I'm adding validation to DecryptionMaterials and EncryptionMaterials in both the setPlaintextDataKey and in the Builder.build methods so that it will be impossible to construct an invalid DecryptionMaterials/EncryptionMaterials

Reply:
Actually I'm going to hold off on renaming this now, since that will break existing clients that have written their own CryptoMaterialsManager. Also I will only add the validation to the new setCleartextDataKey methods

Reply:
Discussed with @seebees, I'm going to try defining new versions of DecryptionMaterials and EncryptionMaterials for exclusive use in Keyrings